scripts/externalTests/test_helpers.py

```diff
@@ -0,0 +1,284 @@
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import os
+import re
+import subprocess
+import sys
+from abc import ABCMeta
+from argparse import ArgumentParser
+from dataclasses import dataclass, field
+from pathlib import Path
+from shutil import rmtree
+from tempfile import mkdtemp
+from textwrap import dedent
+from typing import List
+
+# Our scripts/ is not a proper Python package so we need to modify PYTHONPATH to import from it
+# pragma pylint: disable=import-error,wrong-import-position
+PROJECT_ROOT = Path(__file__).parents[2]
+sys.path.insert(0, f"{PROJECT_ROOT}/scripts/common")
+
+from git_helpers import git_commit_hash
+
+SOLC_FULL_VERSION_REGEX = re.compile(r"^[a-zA-Z: ]*(.*)$")
+SOLC_SHORT_VERSION_REGEX = re.compile(r"^([0-9.]+).*\+|\-$")
+
+evm_version = os.environ.get("DEFAULT_EVM")
+CURRENT_EVM_VERSION: str = evm_version if evm_version is not None else "shanghai"
+
+AVAILABLE_PRESETS: List[str] = [
+    "legacy-no-optimize",
+    "ir-no-optimize",
+    "legacy-optimize-evm-only",
+    "ir-optimize-evm-only",
+    "legacy-optimize-evm+yul",
+    "ir-optimize-evm+yul",
+]
+
+
+@dataclass
+class TestConfig:
+    name: str
+    repo_url: str
+    ref_type: str
+    ref: str
+    build_dependency: str = field(default="nodejs")
+    compile_only_presets: List[str] = field(default_factory=list)
+    settings_presets: List[str] = field(default_factory=lambda: AVAILABLE_PRESETS)
+    evm_version: str = field(default=CURRENT_EVM_VERSION)
+
+    def selected_presets(self):
+        return set(self.compile_only_presets + self.settings_presets)
+
+
+class TestRunner(metaclass=ABCMeta):
+    config: TestConfig
+    solc_binary_type: str
+    solc_binary_path: Path
+
+    def __init__(self, argv, config: TestConfig):
+        args = parse_command_line(f"{config.name} external tests", argv)
+        self.config = config
+        self.solc_binary_type = args.solc_binary_type
+        self.solc_binary_path = args.solc_binary_path
+        self.env = os.environ.copy()
+        self.tmp_dir = mkdtemp(prefix=f"ext-test-{config.name}-")
+        self.test_dir = Path(self.tmp_dir) / "ext"
+
+    def setup_solc(self) -> str:
+        if self.solc_binary_type == "solcjs":
+            # TODO: add support to solc-js # pylint: disable=fixme
+            raise NotImplementedError()
+        print("Setting up solc...")
+        solc_version_output = subprocess.getoutput(f"{self.solc_binary_path} --version").split(":")[1]
+        return parse_solc_version(solc_version_output)
+
+    @staticmethod
+    def on_local_test_dir(fn):
+        """Run a function inside the test directory"""
+
+        def f(self, *args, **kwargs):
+            assert self.test_dir is not None
+            os.chdir(self.test_dir)
+            return fn(self, *args, **kwargs)
+
+        return f
```

Comment:
> I think that putting directory changes in a modifier makes things harder to follow. When you call the function it's not really apparent if it has this modifier applied or not, so this side-effect may lead to surprises. It also does not restore the original directory when the function exits.

Yeah, as all commands should run from the test directory I imagined that it would be useful to have a decorator there, so you could ensure that you are in the correct directory when calling the methods. But indeed, it is not foolproof, and maybe just some asserts are sufficient, as you suggested. And whoever is writing the external tests and overriding the methods, need to ensure that they are in the right directory.

> Would it not be simpler to just have a function called `TestRunner.enter_test_dir()` and explicitly call it once inside `run_test()`?

Maybe. But if some method of the external test change directory, it would need to ensure that the original directory is restored after execute the method. As I currently change to the test directory on every method, using the decorator, I don't need to worry with restoring the original directory.

> If you want to make sure that you won't forget to change dir before invoking these functions, you should instead make them assert that they're in test dir.

Yeah, but I'm not really sure where those asserts should go, since the methods of a runner can be overridden.

Reply:
> Maybe. But if some method of the external test change directory, it would need to ensure that the original directory is restored after execute the method. As I currently change to the test directory on every method, using the decorator, I don't need to worry with restoring the original directory.

Ah, so that's the main concern. I assumed it was mostly for convenience. I guess we can leave the decorator in place, but I still think it should restore the original dir when the function finishes (using a `finally` block). I.e. it should only have effect on the wrapped method and leave things as they were for the code that calls it. This way the behavior is easier to reason about because path changes in `run_test()` and in these methods are independent of each other.

Reply:
Also, I tried to check if there's a pathlib equivalent to `os.chdir()`. Surprisingly, there isn't and it's actually recommended to always specify path explicitly via `cwd` in `subprocess.run()`. That could be an alternative to messing with working dir.

Reply:
I did using `finally` as you suggested :)